### PR TITLE
feat: Query if ScrollViewState is_at_bottom()

### LINKED ISSUE #72: ScrollViewState: Add helper interface
Hello 🐦 

In my app I want the following behavior while using `ScrollViewState` under the hood:
A user runs the app and sees e.g. a log. The log is rendered using `ScrollView`+`ScrollViewState`.
The log regularly gets updates and grows. 
Because I can call `.scroll_to_bottom()` during each render, the user always sees the latest log entries.

Now, the user scrolls up the log and observes older entries. Even though new entries get added to the log, the view remains fixed on the position the user scrolled to.

Later, the user scrolls back down to end of the scroll view, seeing the latest entries again.
When new entries are added, the view remains at the bottom as before.

Here are some options that I believe make `ScrollView` easier to use. Let me know if you have further questions. Should you accept my proposal, I would love to file a PR.

### My current solution

The current solution works fine but feels a bit unclean. I manually keep track on the scroll position, replicating work that is already been done inside `ScrollViewState` iirc.
```Rust
event_loop(..) {
    MouseEventKind::ScrollDown => {
        self.scroll_position = self.scroll_position.saturating_sub(1);
        self.follow_the_bottom = self.scroll_position == 0;
        self.scroll_state.scroll_down();
    }
    MouseEventKind::ScrollUp => {
        self.follow_the_bottom = false;
        if self.scroll_state.offset().y != 0 {
            self.scroll_position = self.scroll_position.saturating_add(1);
        }
        self.scroll_state.scroll_up();
    }
}

/* ... */
fn draw(...) {
    if self.follow_the_bottom {
        self.scroll_state.scroll_to_bottom()
    }
}
```

### Proposed solution

The `ScrollViewState` can just tell me where the scroll position is. 
```Rust
event_loop(..) {
    MouseEventKind::ScrollDown => {
        self.scroll_state.scroll_down();
        self.follow_the_bottom = self.scroll_state.is_at_bottom();
    }
    MouseEventKind::ScrollUp => {
        self.follow_the_bottom = false;
        self.scroll_state.scroll_up();
    }
}

/* ... */
fn draw(...) {
    if self.follow_the_bottom {
        self.scroll_state.scroll_to_bottom()
    }
}
```

### Alternative solution
Let me ask for the size of the scroll view. Theoretically, I could calculate this info myself but in practice, that leads to code in the event loop (or where ever you have your "business" logic) reasoning about the ui layout. It's not too bad but it feels wrong to me.

The `ScrollViewState` can just tell me where the scroll position is. 
```Rust
event_loop(..) {
    MouseEventKind::ScrollDown => {
        self.scroll_state.scroll_down();
        self.follow_the_bottom =  self.scroll_state.offset().y == self.scroll_state.size.height();
    }
    MouseEventKind::ScrollUp => {
        self.follow_the_bottom = false;
        self.scroll_state.scroll_up();
    }
}

/* ... */
fn draw(...) {
    if self.follow_the_bottom {
        self.scroll_state.scroll_to_bottom()
    }
}
```

### PR DESCRIPTION
**Pull request by Teufelchen1**
Hey 🐅,

this aims to fix #72 

It is close to what was discussed but I made the additional decision to respect the page size.
I believe this is the right way to do it as not respecting the page size feels very odd to the user, example:
```Rust
let state = ...;
// our content is 5 rows
assert_eq!(state.size.unwrap().height, 5);
// our page size is 3
assert_eq!(state.page_size.unwrap().height, 3);

// Scrolling down 5 times
state.scroll_down();
state.scroll_down();
state.scroll_down();
state.scroll_down();
state.scroll_down();

// Render it
scroll_view.render(...);

// Are we at the bottom?
assert!(state.is_at_bottom()); // FAILS!
// This is because rendering resets our offset to cater for the page size
assert_eq!(state.offset.unwrap().y, 2);
```
See `scroll_view.rs:210-232` for why that is.

Since that felt wrong to me, I made the judgment call for `is_at_bottom()` to consider the page size. So technically it is now a `are_we_scrolled_down_enough_so_it_behaves_and_looks_like_we_are_at_the_bottom()` 🥸

I added a unit test.

Feedback? :)

### CODE REVIEW

File: tui-scrollview/src/scroll_view.rs [hunk 1]
```diff
@@ -451,6 +450,65 @@ mod tests {
...
+            ])
+        );
+
+        // We could also jump directly to the bottom
+        state.scroll_to_bottom();
+        assert!(state.is_at_bottom());
+
+        // which sets the offset to the last row of content,
+        // ensuring to be at the bottom regardless of the page size.
+        assert_eq!(state.offset.y, state.size.unwrap().height - 1);
```
Comment (Teufelchen1) [inline tui-scrollview/src/scroll_view.rs:509]: More a kind of explanation & documentation rather than a strict unit test.

File: tui-scrollview/src/scroll_view.rs [hunk 2]
```diff
@@ -451,6 +450,65 @@ mod tests {
...
+            buf,
+            Buffer::with_lines(vec![
+                "ABCDE▲",
+                "KLMNO█",
+                "UVWXY█",
+                "EFGHI║",
+                "OPQRS▼",
+                "◄██═► ",
+            ])
+        );
```
[the remark this answers is not in the document]
Comment (joshka) [reply tui-scrollview/src/scroll_view.rs:474]: I tend to prefer not putting these sort of pre-condition assertions that are covered elsewhere in tests. My rationale for this preference is that if some other change fails to render the starting condition, I'd prefer to see this fail at the point where it's testing the end condition, rather than in the pre-conditions. As a general rule, I've found that this makes it easier to debug failing tests while developing as well as to maintain them when things change which would affect pre-conditions but not post-conditions.

File: tui-scrollview/src/scroll_view.rs [hunk 3]
```diff
@@ -787,10 +845,10 @@ mod tests {
         let mut buf = Buffer::empty(Rect::new(0, 0, 7, 5));
         let mut state = ScrollViewState::default();
         let mut list_state = ListState::default();
-        let items: Vec<String> = (1..=10).map(|i| format!("Item {}", i)).collect();
+        let items: Vec<String> = (1..10).map(|i| format!("Item {}", i)).collect();
```
Comment (joshka) [inline tui-scrollview/src/scroll_view.rs:848]: This was intentional. I wanted 10 items, not 9 in the list. What's the purpose of this change?

File: tui-scrollview/src/scroll_view.rs [hunk 4]
```diff
@@ -787,10 +845,10 @@ mod tests {
...
         let list = List::new(items);
         scroll_view.render_stateful_widget(list, scroll_view.area(), &mut list_state);
-        scroll_view.render(buf.area, &mut buf, &mut state);
+        scroll_view.clone().render(buf.area, &mut buf, &mut state);
```
Comment (joshka) [inline tui-scrollview/src/scroll_view.rs:851]: I'm not sure why this clone is needed here

File: tui-scrollview/src/state.rs [hunk 5]
```diff
@@ -82,4 +82,14 @@ impl ScrollViewState {
             .map_or(u16::MAX, |size| size.height.saturating_sub(1));
         self.offset.y = bottom;
     }
+
+    /// True if the scroll view state is at the bottom of the buffer
+    /// Takes the pagesize into account
```
Comment (joshka) [inline tui-scrollview/src/state.rs:87]: nit: Wrap at 100 chars. First line should be a summary sentence followed by a newline then any detailed info.

Here, I think it's worth adding some info on the caveats such as if the size / page size is not yet computed (by rendering), then it won't be taken into account. Likely this info should be also added to scroll_to_bottom docs too.

File: tui-scrollview/src/scroll_view.rs (hunk 3 above)
Comment (Teufelchen1) [reply tui-scrollview/src/scroll_view.rs]: Whoops, that wasn't supposed to be in the commit. Change dropped.

File: tui-scrollview/src/scroll_view.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Teufelchen1) [reply tui-scrollview/src/scroll_view.rs]: Yes, I am with you on that opinion. Dropped it.

File: tui-scrollview/src/state.rs (hunk 5 above)
Comment (Teufelchen1) [reply tui-scrollview/src/state.rs]: I can't wrap at 100 chars as the line is not longer than 100 chars. Or my editor lies to me :shaking_face: 

I added some info on the behavior of both functions. While doing so, I figured it would be best to just define a known behavior in that case. Changing the default for the buffer from `u16:Max` to `0` ensures that the function always returns true in that case. Also spotted a potential panic/undefined behavior, if the scroll position is at `u16:Max` and pagesize is `>= 1`, the computation would overflow the underlying `u16`. Saturated add seems fine to me, objections?

File: tui-scrollview/src/scroll_view.rs (hunk 4 above)
Comment (Teufelchen1) [reply tui-scrollview/src/scroll_view.rs:843]: It is actually reassuring to see that you also didn't get it. I was very confused when Rust hit me with this (when you omit that `clone()`):
```txt
error[E0382]: use of moved value: `scroll_view`
   --> tui-scrollview/src/scroll_view.rs:482:9
    |
454 |     fn move_to_bottom(scroll_view: ScrollView) {
    |                       ----------- move occurs because `scroll_view` has type `scroll_view::ScrollView`, which does not implement the `Copy` trait
...
461 |         scroll_view.render(buf.area, &mut buf, &mut state);
    |                     -------------------------------------- `scroll_view` moved due to this method call
...
482 |         scroll_view.render(buf.area, &mut buf, &mut state);
    |         ^^^^^^^^^^^ value used here after move
    |
note: `ratatui::prelude::StatefulWidget::render` takes ownership of the receiver `self`, which moves `scroll_view`
   --> /home/teufelchen/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ratatui-0.29.0/src/widgets.rs:244:15
    |
244 |     fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State);
    |               ^^^^
help: you can `clone` the value and consume it, but this might not be your desired behavior
    |
461 |         scroll_view.clone().render(buf.area, &mut buf, &mut state);
    |                    ++++++++

For more information about this error, try `rustc --explain E0382`.

```

I have to admit I am unable to explain it really. I figured it has something to do with `#[fixture]` or `#[rstest]` which I don't know about. Hence I just went with the compiler hint / the clone.

File: tui-scrollview/src/state.rs [hunk 6]
```diff
@@ -82,4 +85,17 @@ impl ScrollViewState {
             .map_or(u16::MAX, |size| size.height.saturating_sub(1));
         self.offset.y = bottom;
     }
+
+    /// True if the scroll view state is at the bottom of the buffer
+    ///
+    /// This takes the page size into account. It returns true if the current scroll offset
+    /// plus the page size matches or exceeds the buffer length.
+    /// The buffer and the page size are unkown until computed during the first rendering.
```
Comment (joshka) [inline tui-scrollview/src/state.rs:93]: Typo unkown

File: tui-scrollview/src/state.rs (hunk 5 above)
Comment (Teufelchen1) [reply tui-scrollview/src/state.rs]: > As it stands wrapping with a single line would not cause a new paragraph, so the second line is a continuation of the first. It's wrapped too soon.


Is it correct now?

**Comment by joshka** [comment]
I pushed a rebased/reworked version of this branch to address the review findings.

The two concrete problems found during review were:

1. `is_at_bottom()` reported true one row too early. The old check treated the visible page end as though it were inclusive, but the rendered page end is effectively exclusive. That meant `offset.y + page_size.height == content_height - 1` could report bottom even though the final row was not visible yet.

2. Rendering near the bottom could store the clipped remaining content size as `page_size` instead of the actual scrollbar-adjusted viewport size. That matters because later page scrolling should use the viewport/page size, not the smaller remainder after intersecting with the backing buffer.

The pushed commit sequence is intentionally split so the problem and fix are easier to review:

1. `0d85274b33b7` `test: characterize scrollview bottom edge cases`
   Adds characterization coverage for the existing edge cases before fixing them. The comments in this commit explicitly mark the asserted behavior as incorrect so the failure mode is visible in isolation.

2. `5991b1880db8` `fix: correct scrollview bottom detection`
   Fixes `is_at_bottom()` so the last content row must actually be visible, clamps render offsets against the scrollbar-adjusted viewport size, and records the rendered viewport size as `page_size` before clipping to the content buffer.

3. `e581fe6e43f7` `test: cover scrollview scrollbar visibility policy`
   Adds a direct matrix for `Always` / `Automatic` / `Never` scrollbar visibility combinations. Coverage showed those branches were the weak spot around the viewport-size interaction, so this makes the policy explicit without changing behavior.

Validation run locally:

```sh
cargo +nightly fmt --all
cargo test -p tui-scrollview --all-features
cargo clippy -p tui-scrollview --all-targets --all-features -- -D warnings
cargo llvm-cov -p tui-scrollview --all-features --lcov --output-path /tmp/tui-scrollview-lcov.info
```

**Comment by joshka** [comment]
Thanks for the PR and apologies on the delay in getting to it

**Comment by Teufelchen1** [comment]
Thanks! :heart:
